lib/net/ldap/connection.rb

```diff
@@ -250,131 +250,13 @@ def next_msgid
   def bind(auth)
     instrument "bind.net_ldap_connection" do |payload|
       payload[:method] = meth = auth[:method]
-      if [:simple, :anonymous, :anon].include?(meth)
-        bind_simple auth
-      elsif meth == :sasl
-        bind_sasl(auth)
-      elsif meth == :gss_spnego
-        bind_gss_spnego(auth)
-      else
-        raise Net::LDAP::AuthMethodUnsupportedError, "Unsupported auth method (#{meth})"
-      end
-    end
-  end
-
-  #--
-  # Implements a simple user/psw authentication. Accessed by calling #bind
-  # with a method of :simple or :anonymous.
-  #++
-  def bind_simple(auth)
-    user, psw = if auth[:method] == :simple
-                  [auth[:username] || auth[:dn], auth[:password]]
-                else
-                  ["", ""]
-                end
-
-    raise Net::LDAP::BindingInformationInvalidError, "Invalid binding information" unless (user && psw)
-
-    message_id = next_msgid
-    request    = [
-      LdapVersion.to_ber, user.to_ber,
-      psw.to_ber_contextspecific(0)
-    ].to_ber_appsequence(Net::LDAP::PDU::BindRequest)
-
-    write(request, nil, message_id)
-    pdu = queued_read(message_id)
-
-    if !pdu || pdu.app_tag != Net::LDAP::PDU::BindResult
-      raise Net::LDAP::NoBindResultError, "no bind result"
+      require "net/ldap/auth_adapters/#{meth}"
```

Comment:
This works, but I prefer to have `requires` at the top of a file to make it obvious. Also, until we extract the other adaptors into gems, I'd suggest putting these requires in [`lib/net/ldap.rb`](https://github.com/ruby-ldap/ruby-net-ldap/blob/master/lib/net/ldap.rb#L20-L29):

```ruby
require 'net/ldap/auth_adapter'
require 'net/ldap/auth_adapter/simple'  # <-- side note, can we rename the folder and namespace to be singular?
require 'net/ldap/auth_adapter/sasl'
# ...

# Move all of the registration to a single place, decouples the definition of the adapter from it's registration and use
Net::LDAP::AuthAdapter.register([:simple, :anon, :anonymous], Net::LDAP::AuthAdapters::Simple)
Net::LDAP::AuthAdapter.register(:sasl, Net::LDAP::AuthAdapters::Sasl)
```